src/Processors/QueryPlan/PartsSplitter.cpp

```diff
@@ -94,7 +94,8 @@ std::pair<std::vector<Values>, std::vector<RangesInDataParts>> split(RangesInDat
             parts_ranges_queue.push(
                 {index_access->getValue(part_idx, range.begin), {range, part_idx}, PartsRangesIterator::EventType::RangeStart});
             const auto & index_granularity = parts[part_idx].data_part->index_granularity;
-            if (index_granularity.hasFinalMark() && range.end + 1 == index_granularity.getMarksCount())
+            if (range.end + 1 < index_granularity.getMarksCount()
```

Comment:
previous behaviour was the following. imagine we have a part of 10 marks (1, 10). and we need to read two ranges: (1, 3) and (5, 7). because we didn't track the ends of this ranges we just read the whole part (1, 10) 😱

Reply:
I don't understand why can't we push the end of the range to the `parts_ranges_queue` unconditionally. Otherwise, a comment is needed.